keps/sig-instrumentation/20191115-structured-logging.md

```diff
@@ -0,0 +1,556 @@
+* Broadly adopted by logging libraries with very efficient implementations (zap, zerolog).
+* Out of the box support by many logging backends (Elasticsearch, Stackdriver, BigQuery, Splunk)
+* Easily parsable and transformable
+* Existing tools for ad-hoc analysis (jq)
+
+Example:
+
+```json
+{
+   "ts":"00:15:15.525108",
+   "msg":"Updating ready status",
+   "core/node":{
+      "name":"node-1"
+   },
+   "core/pod":{
+      "name":"nginx-1",
+      "namespace":"default"
+   },
+   "probe":{
+      "ready":false
+   }
+}
+```
+
+### Logging configuration
+
+We would like to introduce new logging configuration options shared by all kubernetes components. `LoggingConfig` structure should be implemented as part of `k8s.io/component-base` options and include common set of flags for logger initialization.
+* `--logging-format` flag should allow to pick between logging interface implementation and allow for backward compatibility. Setting this flag will select particular logger implementation from registry.
+* `-v` flag should allow to change verbosity of logging
+
+Any additional configuration needed by loggers should be provided by flags specific to its implementation. Default loggers should provide a basic customization, but it will not be encouraged to avoid feature bloat. Adding new loggers, changing logger behavior or extending its functionality should be done outside of Kubernetes by implementing logr API and registering it in code.
+
+We acknowledge that extension via code is not the best, but should be sufficient for now.
+
+Proposed flag `--logging-format`  values:
+* `legacy` for old logging format
+* `json` for json format
+
+All logs from the `json` logger should go to standard output.  Out of the existing klog command line arguments only ‘-v’ with be supported in future.
+
+### Kubernetes log schema
+
+Introducing structured logging API is not enough to address one of the biggest issues with log creation or analysis, that is inconsistent metadata. Size of Kubernetes codebase could easily result in drifting naming conventions. As part of this proposal we want to propose a mechanism for ensuring consistency of metadata. To achieve this we will introduce a way to define a schema for subset of commonly used fields in kubernetes. This schema should improve reusability of fields and provide discovery mechanism for logging backend vendors.
+
+Introducing this schema will include three things:
+* Schema definition format
+* Set of helper functions to populate this schema (e.g. `WithObject` for referencing k8s object)
+* Validation mechanism ensuring schema is not bypassed by lower level logging API
+
+Exact schema definition is outside the scope of this proposal and will be addressed by a separate KEP when structured logging effort will enter Beta and logs will start to be migrated. At the end we would expect schema to look similar to [Elastic Common Schema](https://github.com/elastic/ecs/tree/master/schemas)
+
+### Migration
```

Comment:
I think a valuable step0 might be to convert externally-facing libs like client-go, so they no longer demand klog or glog from consumer binaries.